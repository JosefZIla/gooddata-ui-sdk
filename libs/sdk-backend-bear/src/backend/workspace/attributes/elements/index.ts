// (C) 2019-2021 GoodData Corporation
import {
    IElementsQueryFactory,
    IElementsQuery,
    IElementsQueryOptions,
    IElementsQueryResult,
    IAttributeElement,
    IElementsQueryAttributeFilter,
} from "@gooddata/sdk-backend-spi";
import { IMeasure, ObjRef } from "@gooddata/sdk-model";
import invariant from "ts-invariant";

import { BearAuthenticatedCallGuard } from "../../../../types/auth";
import { objRefToUri, getObjectIdFromUri } from "../../../../utils/api";
import { GdcExecuteAFM } from "@gooddata/api-model-bear";
import { LimitingAfmFactory } from "./limitingAfmFactory";

export class BearWorkspaceElements implements IElementsQueryFactory {
    constructor(private readonly authCall: BearAuthenticatedCallGuard, public readonly workspace: string) {}

    public forDisplayForm(ref: ObjRef): IElementsQuery {
        return new BearWorkspaceElementsQuery(this.authCall, ref, this.workspace);
    }
}

class BearWorkspaceElementsQuery implements IElementsQuery {
    private limit: number = 50;
    private offset: number = 0;
    private options: IElementsQueryOptions | undefined;
    private attributeFilters: IElementsQueryAttributeFilter[] | undefined;
    private measures: IMeasure[] | undefined;
    // cached AFM used to apply attributeRef and attributeFilters to the element queries
    private limitingAfm: GdcExecuteAFM.IAfm | undefined;

    // cached value of objectId corresponding to identifier
    private objectId: string | undefined;

    constructor(
        private readonly authCall: BearAuthenticatedCallGuard,
        private readonly displayFormRef: ObjRef,
        private readonly workspace: string,
    ) {}

    public withLimit(limit: number): IElementsQuery {
        invariant(limit > 0, `limit must be a positive number, got: ${limit}`);

        this.limit = limit;

        return this;
    }

    public withOffset(offset: number): IElementsQuery {
        this.offset = offset;
        return this;
    }

    public withAttributeFilters(filters: IElementsQueryAttributeFilter[]): IElementsQuery {
        this.attributeFilters = filters;
        return this;
    }

    public withMeasures(measures: IMeasure[]): IElementsQuery {
        this.measures = measures.length > 0 ? measures : undefined;
        return this;
    }

    public withOptions(options: IElementsQueryOptions): IElementsQuery {
        this.options = options;
        return this;
    }

    public async query(): Promise<IElementsQueryResult> {
        const limitingAfmFactory = new LimitingAfmFactory(this.authCall, this.displayFormRef, this.workspace);

        this.limitingAfm = await limitingAfmFactory.getAfm(this.attributeFilters, this.measures);

        return this.queryWorker(this.offset, this.limit, this.options);
    }

    private async getObjectId(): Promise<string> {
        if (!this.objectId) {
            const uri = await objRefToUri(this.displayFormRef, this.workspace, this.authCall);
            this.objectId = getObjectIdFromUri(uri);
        }

        return this.objectId;
    }

    private async queryWorker(
        offset: number,
        limit: number,
        options: IElementsQueryOptions | undefined,
    ): Promise<IElementsQueryResult> {
        const mergedOptions = { ...options, limit, offset, afm: this.limitingAfm };
        const objectId = await this.getObjectId();
        const data = await this.authCall((sdk) =>
            sdk.md.getValidElements(this.workspace, objectId, mergedOptions),
        );

        const { items, paging } = data.validElements;
        const total = Number.parseInt(paging.total, 10);
        const serverOffset = Number.parseInt(paging.offset, 10);
        const { count } = paging;

        const hasNextPage = serverOffset + count < total;

        const emptyResult: IElementsQueryResult = {
            items: [],
            limit: count,
            offset: total,
            totalCount: total,
            next: () => Promise.resolve(emptyResult),
        };

        return {
            items: items.map((element: { element: IAttributeElement }) => element.element),
            limit: count,
            offset: serverOffset,
            totalCount: total,
            next: hasNextPage
                ? () => this.queryWorker(offset + count, limit, options)
                : () => Promise.resolve(emptyResult),
        };
    }
}
