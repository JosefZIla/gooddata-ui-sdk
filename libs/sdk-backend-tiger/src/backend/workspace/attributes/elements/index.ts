// (C) 2019-2021 GoodData Corporation
import {
    IElementsQueryFactory,
    IElementsQuery,
    IElementsQueryOptions,
    IElementsQueryResult,
    UnexpectedError,
    NotSupported,
    IAttributeElement,
} from "@gooddata/sdk-backend-spi";
import { ObjRef, isIdentifierRef } from "@gooddata/sdk-model";
import invariant from "ts-invariant";
import { TigerAuthenticatedCallGuard } from "../../../../types";

export class TigerWorkspaceElements implements IElementsQueryFactory {
    constructor(private readonly authCall: TigerAuthenticatedCallGuard, public readonly workspace: string) {}

    public forDisplayForm(ref: ObjRef): IElementsQuery {
        return new TigerWorkspaceElementsQuery(this.authCall, ref, this.workspace);
    }
}

class TigerWorkspaceElementsQuery implements IElementsQuery {
    private limit: number = 100;
    private offset: number = 0;
    private options: IElementsQueryOptions | undefined;

    constructor(
        private readonly authCall: TigerAuthenticatedCallGuard,
        private readonly ref: ObjRef,
        private readonly workspace: string,
    ) {}

    public withLimit(limit: number): IElementsQuery {
        invariant(limit > 0, `limit must be a positive number, got: ${limit}`);

        this.limit = limit;

        return this;
    }

    public withOffset(offset: number): IElementsQuery {
        this.offset = offset;
        return this;
    }

    public withAttributeFilters(): IElementsQuery {
        throw new NotSupported("not supported");
    }

    public withMeasures(): IElementsQuery {
        throw new NotSupported("not supported");
    }

    public withOptions(options: IElementsQueryOptions): IElementsQuery {
        this.options = options;
        return this;
    }

    public async query(): Promise<IElementsQueryResult> {
        return this.queryWorker(this.offset, this.limit, this.options);
    }

    private async queryWorker(
        offset: number,
        limit: number,
        options: IElementsQueryOptions | undefined,
    ): Promise<IElementsQueryResult> {
        const { ref } = this;
        if (!isIdentifierRef(ref)) {
            throw new UnexpectedError("Tiger backend does not allow referencing objects by URI");
        }

        const response = await this.authCall((client) => {
            const elementsRequest: Parameters<typeof client.labelElements.processElementsRequest>[0] = {
                ...(options?.complement && { complementFilter: options.complement }),
                ...(options?.filter && { patternFilter: options.filter }),
                ...(options?.order && { sortOrder: options.order === "asc" ? "ASC" : "DESC" }),
                limit,
                offset,
                label: ref.identifier,
                workspaceId: this.workspace,
            };

            return client.labelElements.processElementsRequest(elementsRequest);
        });

        const { paging, elements } = response.data;
        const { count, total, offset: serverOffset } = paging;
        const hasNextPage = serverOffset + count < total;

        const emptyResult: IElementsQueryResult = {
            items: [],
            limit,
            offset,
            totalCount: 0,
            next: () => Promise.resolve(emptyResult),
        };

        return {
            items: elements.map(
                (element): IAttributeElement => ({
                    title: element.title,
                    uri: element.primaryTitle,
                }),
            ),
            limit: count,
            offset: serverOffset,
            totalCount: total,
            next: hasNextPage
                ? () => this.queryWorker(offset + count, limit, options)
                : () => Promise.resolve(emptyResult),
        };
    }
}
