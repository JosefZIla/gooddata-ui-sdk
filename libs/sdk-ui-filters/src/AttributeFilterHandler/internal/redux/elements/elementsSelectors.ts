// (C) 2021-2022 GoodData Corporation
import { IAttributeElement } from "@gooddata/sdk-model";
import { createSelector } from "@reduxjs/toolkit";
import invariant from "ts-invariant";

import { ILoadElementsOptions } from "../../../types";
import { selectState } from "../common/selectors";

/**
 * @internal
 */
export const getElementsByKeys = (keys: string[], elementsMap: Record<string, IAttributeElement>) => {
    return keys.map((key) => {
        const element = elementsMap[key];
        invariant(element, `Unable to map selection to elements - element with key "${key}" is not loaded.`);
        return element;
    });
};

/**
 * @internal
 */
export const selectElementKeys = createSelector(selectState, (state) => state.elements.data ?? []);

/**
 * @internal
 */
export const selectElementsCache = createSelector(selectState, (state) => state.elements.cache);

/**
 * @internal
 */
export const selectElements = createSelector(selectElementKeys, selectElementsCache, getElementsByKeys);

/**
 * @internal
 */
export const selectElementsTotalCount = createSelector(selectState, (state) => state.elements.totalCount);

/**
 * @internal
 */
export const selectElementsTotalCountWithCurrentSettings = createSelector(
    selectState,
    (state) => state.elements.totalCountWithCurrentSettings,
);

/**
 * @internal
 */
export const selectStaticElements = createSelector(selectState, (state) => state.config.staticElements ?? []);

/**
 * @internal
 */
export const selectSearch = createSelector(selectState, (state) => state.elements.currentOptions.search);

/**
 * @internal
 */
export const selectOrder = createSelector(selectState, (state) => state.elements.currentOptions.order);

/**
 * @internal
 */
export const selectLimit = createSelector(selectState, (state) => state.elements.currentOptions.limit);

/**
 * @internal
 */
export const selectOffset = createSelector(selectState, (state) => state.elements.currentOptions.offset);

/**
 * @internal
 */
export const selectLimitingAttributeFilters = createSelector(
    selectState,
    (state) => state.elements.currentOptions.limitingAttributeFilters,
);

/**
 * @internal
 */
export const selectLimitingMeasures = createSelector(
    selectState,
    (state) => state.elements.currentOptions.limitingMeasures,
);

/**
 * @internal
 */
export const selectLimitingDateFilters = createSelector(
    selectState,
    (state) => state.elements.currentOptions.limitingDateFilters,
);

/**
 * @internal
 */
export const selectLoadElementsOptions = createSelector(
    selectOffset,
    selectLimit,
    selectOrder,
    selectSearch,
    selectLimitingAttributeFilters,
    selectLimitingMeasures,
    selectLimitingDateFilters,
    (
        offset,
        limit,
        order,
        search,
        limitingAttributeFilters,
        limitingMeasures,
        limitingDateFilters,
    ): ILoadElementsOptions => {
        return {
            limit,
            limitingAttributeFilters,
            limitingDateFilters,
            limitingMeasures,
            offset,
            order,
            search,
        };
    },
);

/**
 * @internal
 */
export const selectLastLoadedElementsOptions = createSelector(selectState, (state): ILoadElementsOptions => {
    return state.elements.lastLoadedOptions;
});
