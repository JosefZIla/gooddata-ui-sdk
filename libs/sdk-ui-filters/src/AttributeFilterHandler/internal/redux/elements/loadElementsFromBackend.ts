// (C) 2022 GoodData Corporation
import isEmpty from "lodash/isEmpty";
import {
    IElementsQueryAttributeFilter,
    IElementsQueryOptions,
    IElementsQueryResult,
} from "@gooddata/sdk-backend-spi";
import {
    attributeElementsIsEmpty,
    IAttributeElements,
    IAttributeMetadataObject,
    newNegativeAttributeFilter,
    ObjRef,
} from "@gooddata/sdk-model";
import { convertError } from "@gooddata/sdk-ui";

import { ILoadElementsOptions } from "../../../types";
import { AttributeFilterHandlerStoreContext } from "../store/types";
import { IHiddenElementsInfo } from "./types";

/**
 * @internal
 */
export function loadElementsFromBackend(
    context: AttributeFilterHandlerStoreContext,
    options: ILoadElementsOptions & { displayFormRef: ObjRef },
    hiddenElementsInfo: IHiddenElementsInfo,
): Promise<IElementsQueryResult> {
    const { backend, workspace } = context;
    const {
        displayFormRef,
        elements,
        limit,
        limitingAttributeFilters,
        limitingDateFilters,
        limitingMeasures,
        offset,
        search,
        order,
        includeTotalCountWithoutFilters,
    } = options;

    let loader = backend.workspace(workspace).attributes().elements().forDisplayForm(displayFormRef);
    const loaderOptions: IElementsQueryOptions = {};

    if (limit) {
        loader = loader.withLimit(limit);
    }
    if (offset) {
        loader = loader.withOffset(offset);
    }
    if (search) {
        loaderOptions.filter = search;
    }
    if (elements) {
        loaderOptions.elements = elements;
    }
    if (order) {
        loaderOptions.order = order;
    }
    if (includeTotalCountWithoutFilters) {
        loaderOptions.includeTotalCountWithoutFilters = includeTotalCountWithoutFilters;
    }
    if (!isEmpty(loaderOptions)) {
        loader = loader.withOptions(loaderOptions);
    }

    if (limitingDateFilters) {
        loader = loader.withDateFilters(limitingDateFilters);
    }

    const effectiveLimitingAttributeFilters = getLimitingAttributeFilters(
        displayFormRef,
        limitingAttributeFilters,
        hiddenElementsInfo.hiddenElements,
        hiddenElementsInfo.attribute,
    );

    if (effectiveLimitingAttributeFilters.length) {
        loader = loader.withAttributeFilters(effectiveLimitingAttributeFilters);
    }

    if (limitingMeasures) {
        loader = loader.withMeasures(limitingMeasures);
    }

    return loader.query().catch((error) => {
        throw convertError(error);
    });
}

function getLimitingAttributeFilters(
    displayFormRef: ObjRef,
    limitingAttributeFilters: IElementsQueryAttributeFilter[] | undefined,
    hiddenElements: IAttributeElements,
    attribute: IAttributeMetadataObject,
): IElementsQueryAttributeFilter[] {
    if (attributeElementsIsEmpty(hiddenElements)) {
        return limitingAttributeFilters ?? [];
    }

    // add a virtual parent filter to get rid of the hidden elements
    // this is the only way that does not mess up offsets since it is handled by the server "natively"
    const hiddenElementsVirtualFilter: IElementsQueryAttributeFilter = {
        attributeFilter: newNegativeAttributeFilter(displayFormRef, hiddenElements),
        overAttribute: attribute.ref,
    };

    return [hiddenElementsVirtualFilter, ...(limitingAttributeFilters ?? [])];
}
