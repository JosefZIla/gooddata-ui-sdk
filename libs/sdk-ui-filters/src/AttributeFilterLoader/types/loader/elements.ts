// (C) 2022 GoodData Corporation
import { IElementsQueryAttributeFilter } from "@gooddata/sdk-backend-spi";
import { IAttributeElement, IMeasure, IRelativeDateFilter } from "@gooddata/sdk-model";
import { IElementsLoadResult, LoadableStatus, CallbackRegistration, Correlation } from "../common";

/**
 * Handles the loading of the elements
 * @alpha
 */
export interface IAttributeElementLoader {
    //
    // manipulators
    //

    /**
     * Trigger the load of a attribute elements range.
     *
     * @remarks
     * Will cancel any running loads if there are any.
     *
     * You can provide a correlation value that will be included in all the events fired by this.
     * This is useful if you want to "pair" loading and loaded events from the same initiated by the same
     * loadElementsRange call.
     *
     * @param offset - the number of elements to skip
     * @param limit - the number of elements to load
     * @param correlation - the correlation value
     */
    loadElementsRange(offset: number, limit: number, correlation?: Correlation): void;

    /**
     * Set the search value used to filter the elements.
     *
     * @remarks
     * MUST NOT trigger a page load. MUST reset any loaded elements as they are no longer relevant.
     *
     * @param search - the search string to use. Use empty string to reset search.
     */
    setSearch(search: string): void;

    /**
     * Set the measure that will limit the available elements.
     *
     * @param measures - the measures to use
     */
    setLimitingMeasures(measures: IMeasure[]): void;

    /**
     * Set the attribute filters that will limit the available elements.
     *
     * @param filters - the filters to use
     */
    setLimitingAttributeFilters(filters: IElementsQueryAttributeFilter[]): void;

    /**
     * Set the date filters that will limit the available elements.
     *
     * @param filters - the filters to use
     */
    setLimitingDateFilters(filters: IRelativeDateFilter[]): void;

    /**
     * Cancel any loading of the elements if any is in progress.
     */
    cancelElementLoad(): void;

    //
    // selectors
    //
    getAllItems(): IAttributeElement[];
    getItemsByKey(keys: string[]): IAttributeElement[];
    getSearch(): string;
    getTotalCount(): number;
    getCountWithCurrentSettings(): number;
    getLoadingStatus(): LoadableStatus;

    //
    // callbacks
    //
    onElementsRangeLoadStart: CallbackRegistration;
    onElementsRangeLoadSuccess: CallbackRegistration<IElementsLoadResult>;
    onElementsRangeLoadError: CallbackRegistration<{ error: Error }>;
    onElementsRangeLoadCancel: CallbackRegistration;
}
