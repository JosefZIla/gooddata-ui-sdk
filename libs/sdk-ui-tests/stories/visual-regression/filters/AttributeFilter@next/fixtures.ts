// (C) 2022 GoodData Corporation

export const ElementUris = {
    CompuSci: "/gdc/md/referenceworkspace/obj/1054/elements?id=165678",
    Educationly: "/gdc/md/referenceworkspace/obj/1054/elements?id=165847",
    Explorer: "/gdc/md/referenceworkspace/obj/1054/elements?id=342345",
    "Grammar Plus": "/gdc/md/referenceworkspace/obj/1054/elements?id=168737",
    PhoenixSoft: "/gdc/md/referenceworkspace/obj/1054/elements?id=166339",
    TouchAll: "/gdc/md/referenceworkspace/obj/1054/elements?id=460491",
    WonderKid: "/gdc/md/referenceworkspace/obj/1054/elements?id=166497",
};
