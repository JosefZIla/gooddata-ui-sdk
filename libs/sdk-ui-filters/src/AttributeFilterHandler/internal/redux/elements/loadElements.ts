// (C) 2022 GoodData Corporation
import { IElementsQueryResult } from "@gooddata/sdk-backend-spi";
import { IAttributeElement, ObjRef } from "@gooddata/sdk-model";

import { ILoadElementsOptions } from "../../../types";
import { AttributeFilterHandlerStoreContext } from "../store/types";
import { IHiddenElementsInfo } from "./types";
import { loadElementsFromStaticElements } from "./loadElementsFromStaticElements";
import { loadElementsFromBackend } from "./loadElementsFromBackend";

/**
 * @internal
 */
export async function loadElements(
    context: AttributeFilterHandlerStoreContext,
    options: ILoadElementsOptions & { displayFormRef: ObjRef },
    hiddenElementsInfo: IHiddenElementsInfo,
    staticElements: IAttributeElement[] | undefined,
): Promise<IElementsQueryResult> {
    return staticElements?.length
        ? loadElementsFromStaticElements(options, hiddenElementsInfo, staticElements)
        : loadElementsFromBackend(context, options, hiddenElementsInfo);
}
