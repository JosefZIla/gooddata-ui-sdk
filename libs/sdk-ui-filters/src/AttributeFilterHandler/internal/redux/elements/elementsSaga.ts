// (C) 2022 GoodData Corporation
import { SagaIterator } from "redux-saga";
import { call, select, SagaReturnType } from "redux-saga/effects";

import { ILoadElementsResult, ILoadElementsOptions } from "../../../types";
import {
    selectAttributeFilterDisplayForm,
    selectHiddenElementsAsAttributeElements,
} from "../filter/filterSelectors";
import { selectAttribute } from "../loadAttribute/loadAttributeSelectors";
import { getAttributeFilterContext, PromiseFnReturnType } from "../common/sagas";
import { selectStaticElements } from "./elementsSelectors";
import { loadElements } from "./loadElements";

/**
 * @internal
 */
export function* elementsSaga(options: ILoadElementsOptions): SagaIterator<ILoadElementsResult> {
    const context: SagaReturnType<typeof getAttributeFilterContext> = yield call(getAttributeFilterContext);

    const hiddenElements: ReturnType<typeof selectHiddenElementsAsAttributeElements> = yield select(
        selectHiddenElementsAsAttributeElements,
    );

    const attribute: ReturnType<typeof selectAttribute> = yield select(selectAttribute);
    const attributeFilterDisplayFormRef: ReturnType<typeof selectAttributeFilterDisplayForm> = yield select(
        selectAttributeFilterDisplayForm,
    );

    const staticElements: ReturnType<typeof selectStaticElements> = yield select(selectStaticElements);

    const elementsQueryResult: PromiseFnReturnType<typeof loadElements> = yield call(
        loadElements,
        context,
        {
            displayFormRef: attributeFilterDisplayFormRef,
            ...options,
        },
        {
            hiddenElements,
            attribute,
        },
        staticElements,
    );

    return {
        elements: elementsQueryResult.items,
        totalCount: elementsQueryResult.totalCount,
        options,
    };
}
