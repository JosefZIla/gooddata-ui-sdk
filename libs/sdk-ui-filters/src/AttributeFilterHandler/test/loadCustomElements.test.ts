// (C) 2019-2022 GoodData Corporation
import { newTestAttributeFilterHandler } from "./fixtures";
import { waitForAsync } from "./testUtils";
import * as elements from "../internal/redux/elements/loadElements";
import { BadRequestSdkError } from "@gooddata/sdk-ui";

describe("AttributeFilterHandler", () => {
    it("loadCustomElements() should trigger onLoadCustomElementsStart() callback", async () => {
        const onLoadCustomElementsStart = jest.fn();
        const attributeFilterHandler = newTestAttributeFilterHandler("positive");

        attributeFilterHandler.init();
        await waitForAsync();

        attributeFilterHandler.onLoadCustomElementsStart(onLoadCustomElementsStart);
        attributeFilterHandler.loadCustomElements({}, "start");

        await waitForAsync();

        expect(onLoadCustomElementsStart).toHaveBeenCalledTimes(1);
        expect(onLoadCustomElementsStart.mock.calls[0]).toMatchSnapshot("with parameters");
    });

    it("loadCustomElements() that was successful should trigger onLoadCustomElementsSuccess() callback", async () => {
        const onLoadCustomElementsSuccess = jest.fn();
        const attributeFilterHandler = newTestAttributeFilterHandler("positive");

        attributeFilterHandler.init();
        await waitForAsync();

        attributeFilterHandler.onLoadCustomElementsSuccess(onLoadCustomElementsSuccess);
        attributeFilterHandler.loadCustomElements({}, "success");

        await waitForAsync();

        expect(onLoadCustomElementsSuccess).toHaveBeenCalledTimes(1);
        expect(onLoadCustomElementsSuccess.mock.calls[0]).toMatchSnapshot("with parameters");
    });

    it("loadCustomElements() that failed should trigger onLoadCustomElementsError() callback", async () => {
        const onLoadCustomElementsError = jest.fn();
        const attributeFilterHandler = newTestAttributeFilterHandler("positive");

        attributeFilterHandler.init();
        await waitForAsync();

        jest.spyOn(elements, "loadElements").mockRejectedValueOnce(new BadRequestSdkError("Elements error"));
        attributeFilterHandler.onLoadCustomElementsError(onLoadCustomElementsError);
        attributeFilterHandler.loadCustomElements({}, "error");

        await waitForAsync();

        expect(onLoadCustomElementsError).toHaveBeenCalledTimes(1);
        expect(onLoadCustomElementsError.mock.calls[0]).toMatchSnapshot("with parameters");
    });

    it("loadCustomElements() that was canceled by cancelCustomElementsLoad() call should trigger onLoadCustomElementsCancel() callback", async () => {
        const onLoadCustomElementsCancel = jest.fn();
        const attributeFilterHandler = newTestAttributeFilterHandler("positive");

        attributeFilterHandler.init();
        await waitForAsync();

        attributeFilterHandler.onLoadCustomElementsCancel(onLoadCustomElementsCancel);
        attributeFilterHandler.loadCustomElements({}, "cancel");
        attributeFilterHandler.cancelCustomElementsLoad("cancel");
        await waitForAsync();

        expect(onLoadCustomElementsCancel).toHaveBeenCalledTimes(1);
        expect(onLoadCustomElementsCancel.mock.calls[0]).toMatchSnapshot("with parameters");
    });

    it("loadCustomElements() that was called multiple times should trigger onLoadCustomElementsStart() callback multiple times", async () => {
        const onLoadCustomElementsStart = jest.fn();
        const attributeFilterHandler = newTestAttributeFilterHandler("positive");

        attributeFilterHandler.init();
        await waitForAsync();

        attributeFilterHandler.onLoadCustomElementsStart(onLoadCustomElementsStart);
        attributeFilterHandler.loadCustomElements({}, "1");
        attributeFilterHandler.loadCustomElements({}, "2");

        await waitForAsync();

        expect(onLoadCustomElementsStart).toHaveBeenCalledTimes(2);
    });

    it("loadCustomElements() that was called multiple times should trigger onLoadCustomElementsSuccess() callback multiple times", async () => {
        const onLoadCustomElementsSuccess = jest.fn();
        const attributeFilterHandler = newTestAttributeFilterHandler("positive");

        attributeFilterHandler.init();
        await waitForAsync();

        attributeFilterHandler.onLoadCustomElementsSuccess(onLoadCustomElementsSuccess);
        attributeFilterHandler.loadCustomElements({}, "1");
        attributeFilterHandler.loadCustomElements({}, "2");

        await waitForAsync();

        expect(onLoadCustomElementsSuccess).toHaveBeenCalledTimes(2);
    });

    it("loadCustomElements() that was called multiple times should trigger onLoadCustomElementsError() callback multiple times", async () => {
        const onLoadCustomElementsError = jest.fn();
        const attributeFilterHandler = newTestAttributeFilterHandler("positive");

        attributeFilterHandler.init();
        await waitForAsync();

        attributeFilterHandler.onLoadCustomElementsError(onLoadCustomElementsError);
        jest.spyOn(elements, "loadElements").mockRejectedValueOnce(new BadRequestSdkError("Elements error"));
        attributeFilterHandler.loadCustomElements({}, "1");
        jest.spyOn(elements, "loadElements").mockRejectedValueOnce(new BadRequestSdkError("Elements error"));
        attributeFilterHandler.loadCustomElements({}, "2");

        await waitForAsync();

        expect(onLoadCustomElementsError).toHaveBeenCalledTimes(2);
    });
});
