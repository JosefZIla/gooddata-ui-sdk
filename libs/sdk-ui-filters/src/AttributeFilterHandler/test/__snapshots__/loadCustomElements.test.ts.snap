// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttributeFilterHandler loadCustomElements() should trigger onLoadCustomElementsStart() callback: with parameters 1`] = `
Array [
  Object {
    "correlation": "start",
  },
]
`;

exports[`AttributeFilterHandler loadCustomElements() that failed should trigger onLoadCustomElementsError() callback: with parameters 1`] = `
Array [
  Object {
    "correlation": "error",
    "error": [Error: Elements error],
  },
]
`;

exports[`AttributeFilterHandler loadCustomElements() that was canceled by cancelCustomElementsLoad() call should trigger onLoadCustomElementsCancel() callback: with parameters 1`] = `
Array [
  Object {
    "correlation": "cancel",
  },
]
`;

exports[`AttributeFilterHandler loadCustomElements() that was successful should trigger onLoadCustomElementsSuccess() callback: with parameters 1`] = `
Array [
  Object {
    "correlation": "success",
    "elements": Array [
      Object {
        "title": "CompuSci",
        "uri": "/gdc/md/referenceworkspace/obj/1054/elements?id=165678",
      },
      Object {
        "title": "Educationly",
        "uri": "/gdc/md/referenceworkspace/obj/1054/elements?id=165847",
      },
      Object {
        "title": "Explorer",
        "uri": "/gdc/md/referenceworkspace/obj/1054/elements?id=342345",
      },
      Object {
        "title": "Grammar Plus",
        "uri": "/gdc/md/referenceworkspace/obj/1054/elements?id=168737",
      },
      Object {
        "title": "PhoenixSoft",
        "uri": "/gdc/md/referenceworkspace/obj/1054/elements?id=166339",
      },
      Object {
        "title": "TouchAll",
        "uri": "/gdc/md/referenceworkspace/obj/1054/elements?id=460491",
      },
      Object {
        "title": "WonderKid",
        "uri": "/gdc/md/referenceworkspace/obj/1054/elements?id=166497",
      },
    ],
    "options": Object {},
    "totalCount": 7,
  },
]
`;
